bundles/org.openhab.core.automation.module.script.rulesupport/src/main/java/org/openhab/core/automation/module/script/rulesupport/loader/AbstractScriptFileWatcher.java

```diff
@@ -202,158 +206,157 @@ protected boolean watchSubDirectories() {
-        assert scriptType.isPresent();
+    private void importFileWhenReady(String scriptIdentifier) {
+        scheduler.submit(() -> {
+            ScriptFileReference ref = scriptMap.get(scriptIdentifier);
+            if (ref == null) {
+                logger.warn("Failed to import script '{}': script reference not found", scriptIdentifier);
+                return;
+            }
 
-        try (InputStreamReader reader = new InputStreamReader(
-                new BufferedInputStream(ref.getScriptFileURL().openStream()), StandardCharsets.UTF_8)) {
-            logger.info("Loading script '{}'", fileName);
+            ref.getQueueStatus().set(false);
 
-            String scriptIdentifier = ref.getScriptIdentifier();
+            Lock lock = getLockForScript(scriptIdentifier);
+            try {
+                if (ref.isLoaded()) {
+                    manager.removeEngine(scriptIdentifier);
+                    logger.debug("Unloaded script '{}'", scriptIdentifier);
+                }
 
-            ScriptEngineContainer container = manager.createScriptEngine(scriptType.get(), scriptIdentifier);
+                if (manager.isSupported(ref.getScriptType()) && ref.getStartLevel() <= currentStartLevel) {
+                    logger.info("(Re-)Loading script '{}'", scriptIdentifier);
+                    if (createAndLoad(ref)) {
+                        ref.setStatus(Status.LOADED);
+                    } else {
+                        // make sure script engine is successfully closed and the loading is re-tried
+                        manager.removeEngine(ref.getScriptIdentifier());
+                        ref.setStatus(Status.PENDING);
+                    }
+                } else {
+                    ref.setStatus(Status.PENDING);
+                    logger.debug("Enqueued script '{}'", ref.getScriptIdentifier());
+                }
+            } finally {
+                lock.unlock();
+            }
+        });
+    }
 
-            if (container != null) {
-                container.getScriptEngine().put(ScriptEngine.FILENAME, fileName);
-                manager.loadScript(container.getIdentifier(), reader);
+    private boolean createAndLoad(ScriptFileReference ref) {
+        String scriptIdentifier = ref.getScriptIdentifier();
 
-                logger.debug("Script loaded: {}", fileName);
-                return true;
+        try (InputStreamReader reader = new InputStreamReader(Files.newInputStream(ref.getScriptFilePath()),
+                StandardCharsets.UTF_8)) {
+            ScriptEngineContainer container = manager.createScriptEngine(ref.getScriptType(),
+                    ref.getScriptIdentifier());
+            if (container != null) {
+                container.getScriptEngine().put(ScriptEngine.FILENAME, scriptIdentifier);
+                if (manager.loadScript(container.getIdentifier(), reader)) {
+                    return true;
+                }
```

Comment:
@J-N-K I am wondering whether the "Script loading error, ignoring file" log message shouldn't be here in an `else` statement, because the `else` statament of `container != null` will get called when creation of a script engine fails.
The actual loading of a script fails inside the container's null check if.

Reply:
I think the `else` around the logging should be removed. The loading fails in both cases: `container==null` or script error.

Reply:
Better solution, thanks :+1: